### ISSUE TITLE
Possible issue in the revision codes document

### ISSUE THREAD

**Issue by jriwanek** [raspberrypi/documentation#1072]
revision 0010 (Sony UK 1B+) says the board revision is 1.0, but the silk screen on the two boards I have access to say it's 1.2, is the data table inaccurate or is there something else going on?

**Comment by jriwanek** [comment]
A revision 0014 CM1 also claims to be revision 1.1, wheres the documentation says 1.0

**Comment by JamesH65** [comment]
Are you talking about the table on this page? https://www.raspberrypi.org/documentation/hardware/raspberrypi/revision-codes/README.md

**Comment by JamesH65** [comment]
How are you getting the revision numbers from the boards - those are from the old naming scheme, so would imply very old boards.

**Comment by bennuttall** [comment]
I think @ghollingworth knows all the revision codes by heart, can he help?

**Comment by JamesH65** [comment]
Spoken to he who must be believed. That revision number in the table does not necessarily refer to the revision number on the PCB, but is the first release revision of the particular HW. It is confusing, and is one of the reasons we have moved to the much more sophisticated naming scheme. All current product should be using the new scheme.

In precis, nothing to worry about, just an idiosyncracy of the old naming scheme.

**Comment by jriwanek** [comment]
@JamesH65 `vcgencmd otp_dump` then comparing against the list at <https://github.com/raspberrypi/documentation/blob/master/hardware/raspberrypi/revision-codes/README.md>, So, I guess it's "x version or newer", right?

**Comment by JamesH65** [comment]
`cat /proc/cpuinfo` will report the revision code more easily. 

The new revision scheme came in with the Pi2, so if you are getting old revision codes then the device must be older than that. 

x version or newer is a good way of looking at it.

**Comment by jriwanek** [comment]
I don't suppose there's a more accurate way of getting the *actual* HW revision of the board for the old revision codes is there? I'm working on a python script which gives detailed information for the pi's, I'm assuming there probably is as there's stuff in the firmware for different HW revisions (bplus1, bplus2, 2b1, 2b2, 3b1, 3b2 and so on), but being able to query it to get the exact revision would be really useful.

**Comment by bennuttall** [comment]
> I'm working on a python script which gives detailed information for the pi's

There's a comprehensive one in gpiozero. See [data.py](https://github.com/RPi-Distro/python-gpiozero/blob/master/gpiozero/pins/data.py), [pinout docs](https://gpiozero.readthedocs.io/en/latest/cli_tools.html) and [pi_info docs](https://gpiozero.readthedocs.io/en/latest/api_info.html). Install from github master for the most up-to-date version. Anything you're missing?

**Comment by jriwanek** [comment]
Looks like you cover a few things I don't but you don't *appear* to cover a few of the things I'd like to, for example, the `bplus1` revision has Power Low on GPIO 31, but on the `bplus2` it's on 35, with `LAN_RUN` being on the opposite.

**Comment by jriwanek** [comment]
I also note that your manufacturer info differs from the official docs for 000f and 0013.

**Comment by jriwanek** [comment]
Completeness is the name of the game, yes; any comment on whether the MFR for the 000f is Egoman or Qisda, and whether the 0013 is Embest or Egoman?

**Comment by jriwanek** [comment]
Also, I'm guessing 2b1 in the firmware maps to the 2B 1.0 board, 2b1 maps to 2B 1.1, and 2B 1.2 shares with the 3b?

Also, what about `3b1` and `3b2`?

**Comment by jriwanek** [comment]
Oh, and while there is certainly overlap with pizero, the script I'm making has a different audience and intentions: <https://github.com/jriwanek/RaspberryPi/blob/master/OTPParser.py>

**Comment by JamesH65** [comment]
Could be either. I doubt the table is exclusive, quite possible that it states the main manufacturer, and that there were also devices made in other places. But it was up to 6 years ago, so might be difficult to find out. for sure, and it's really not worth spending any time on at this end.

Not sure what you mean with your other questions. There's only one basic firmware, it runs on all boards, and is backwards compatible.  We do provide different versions for different purposes (start_x.elf, start.elf etc), but under the skin they are the same just with added features for this like camera support.  https://www.raspberrypi.org/documentation/configuration/boot_folder.md

**Comment by jriwanek** [comment]
If you look in the start.elf, there's a pin setup table, with different info for different submodels.

**Comment by JamesH65** [comment]
That's device tree stuff. There are usually different device trees for different models, but the end result is that on the whole, all models behave the same way, or as close as they can be given different HW features.

**Comment by jriwanek** [comment]
Yeah, I'm aware, But being able to map to the list for said requires accurate version information, which is why I'm asking.

**Comment by jriwanek** [comment]
Let me put it another way: Is there a way for me to query the exact board revision from the ARM side in cases where the official table doesn't cover all possibilities, or am I going to have to make a custom firmware to grok it from the VC4 to get info like which sub-revision a board is like with the 2 possibilities for the B Plus, the 2B and the 3B?

**Comment by JamesH65** [comment]
Just taken a look at the firmware which checks all this stuff, and I still cannot answer your question, its a bit convoluted. The old style revision code is morphed in to a new style one which is then, I think, used to select the correct DT. The boards revison numbers are hardcoded as the old style code does not provide that information.

**Comment by lurch** [comment]
@pelwell I think what's being asked for, is if it's possible from within Linux to tell which device-tree got loaded by the firmware?

> for example, the bplus1 revision has Power Low on GPIO 31, but on the bplus2 it's on 35, with LAN_RUN being on the opposite.

Oooh, that's interesting. @bennuttall I guess that means that https://gpiozero.readthedocs.io/en/stable/recipes_advanced.html#controlling-the-pi-s-own-leds might need an additional note?

**Comment by jriwanek** [comment]
@JamesH65 The integrated DT blob specifically, not the standalone one found in /boot on linux. (I already know how to query that using vcdbg)

**Comment by lurch** [comment]
"integrated DT blob"?? Is https://github.com/raspberrypi/documentation/blob/master/configuration/pin-configuration.md what you're looking for?

**Comment by jriwanek** [comment]
No. In the firmware itself, there's a small blob which sets stuff like CAM_SHUTDOWN and ETH_CLK, it's what does the initial pin configuration, before the external blob gets loaded if you're using a dt-blob.bin.

**Comment by JamesH65** [comment]
TBH, spent too much time on this already. Niche use case. i.e. just you, and since I don't know the answers to any of the questions, I'm bowing out.

**Comment by jriwanek** [comment]
Thanks anyways James, I'll muddle through and figure it out if nobody else has suggestions.

**Comment by bennuttall** [comment]
Yeah this is way off topic, best take it to the forums. And if you have any suggestions for gpiozero open an issue there.

**Comment by jriwanek** [comment]
Will do ben.

**Comment by lurch** [comment]
Did you read all the way to the bottom of the page? :wink: 
https://github.com/raspberrypi/documentation/blob/master/configuration/pin-configuration.md#sample-device-tree-source-file  "It is the master Raspberry Pi blob"

**Comment by jriwanek** [comment]
Yes, but that doesn't tell me how to map which block is currently in use for the afformentioned cases.

**Comment by pelwell** [comment]
TL;DR - the documentation is inaccurate.

@JamesH65 If you think that's convoluted, you should see what it used to look like.

@lurch Please don't confuse dt-blob with Device Tree - users find it hard enough as it is.

@jriwanek Yes, you are correct - the table is wrong for 0x10 which should appear as a rev 1.2 B+, and 0x14 is a rev 1.0 CM. The firmware should decode it as such (`cat /proc/device-tree/model`).
The simplest, least invasive change is to make the sanitised internal board revision number available to applications, which can then pick out the fields in the usual way.

**Comment by jriwanek** [comment]
Thank you so much Phil!

**Comment by lurch** [comment]
Ooops, sorry if I caused any confusion :-/

**Comment by jriwanek** [comment]
Well, we got there in the end, so no harm, no foul lurch.
